src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs

```diff
@@ -12,8 +12,28 @@ public class Publishers
     {
         internal IEnumerable<PublisherAddress> GetPublisherFor(Type eventType)
         {
-            var distinctPublishers = rules.Select(r => r.Apply(eventType)).Where(e => e != null).Distinct().ToList();
-            return distinctPublishers;
+            List<PublisherAddress> staticPublishersForType;
+            if (!staticPublishers.TryGetValue(eventType, out staticPublishersForType))
+            {
+                staticPublishersForType = emptyList;
+            }
+
+            if (dynamicRules.Count > 0)
+            {
+                var dynamicAddresses = new List<PublisherAddress>();
+                foreach (var rule in dynamicRules)
+                {
+                    var address = rule(eventType);
+                    if (address != null)
+                    {
+                        dynamicAddresses.Add(address);
+                    }
+                }
+
+                return staticPublishersForType.Count > 0 ? staticPublishersForType.Concat(dynamicAddresses) : dynamicAddresses;
+            }
+
+            return staticPublishersForType;
```

Comment:
(also, keeping the current implementation follows the same behavior as before the changes, since the distinct call here was basically useless as you saw)